File: leetcode/easy/HappyNumber.py
class Solution:
    def _sum_squares(self, num: int) -> int:
        """Helper function to calculate the sum of the squares of the digits."""
        total_sum = 0
        while num > 0:
            # divmod is a clean way to get both quotient and remainder
            num, digit = divmod(num, 10)
            total_sum += digit ** 2
        return total_sum

    def isHappy(self, n: int) -> bool:
        """
        Determines if a number is a "happy number".

        A happy number is a number which eventually reaches 1 when replaced by the
        sum of the square of its digits. If the process results in an endless
        cycle of numbers that does not include 1, the number is not happy.

        This solution uses Floyd's Cycle-Finding Algorithm (also known as the
        "Tortoise and Hare" algorithm) to detect cycles with O(1) space complexity,
        which is an optimization over using a hash set.

        Algorithm:
        1. Use two pointers, `slow` and `fast`.
        2. `slow` moves one step at a time through the sequence (n -> next_n).
        3. `fast` moves two steps at a time (n -> next_n -> next_next_n).
        4. If `n` is a happy number, the `fast` pointer will reach 1.
        5. If `n` is not a happy number, it will enter a cycle. In this case,
           the `fast` pointer will eventually "lap" the `slow` pointer, and
           they will become equal.

        Time Complexity: O(log n) - The number of steps is related to the number of digits.
        Space Complexity: O(1) - No extra storage is needed besides the pointers.
        """
        slow = n
        fast = self._sum_squares(n)

        # The loop terminates if a cycle is found (slow == fast)
        # or if the sequence reaches 1.
        while fast != 1 and slow != fast:
            slow = self._sum_squares(slow)
            fast = self._sum_squares(self._sum_squares(fast))

        # If the loop ended because fast reached 1, it's a happy number.
        return fast == 1

File: leetcode/easy/HappyNumber.java
/**
 * Solution class for the Happy Number problem.
 */
class Solution {
    /**
     * Determines if a number is a "happy number".
     * A happy number is a number which eventually reaches 1 when replaced by the sum of the square of its digits.
     * If the process enters a cycle that does not include 1, the number is not happy.
     *
     * This solution uses Floyd's Cycle-Finding Algorithm (the "tortoise and the hare" algorithm)
     * to detect cycles in the sequence of numbers generated.
     *
     * @param n The number to check.
     * @return true if n is a happy number, false otherwise.
     */
    public boolean isHappy(int n) {
        // Initialize two pointers, a slow one and a fast one, for cycle detection.
        int slow = n;
        // The fast pointer moves two steps for every one step of the slow pointer.
        int fast = getNext(n);

        // The loop continues as long as the sequence hasn't reached 1 and the pointers haven't met.
        // If fast == 1, the number is happy.
        // If slow == fast, a cycle has been detected.
        while (fast != 1 && slow != fast) {
            slow = getNext(slow);          // Slow pointer moves one step.
            fast = getNext(getNext(fast)); // Fast pointer moves two steps.
        }

        // If the loop ended, we check why. If fast is 1, we reached the happy state.
        // If the loop ended because slow == fast (and fast is not 1), it's an unhappy cycle.
        return fast == 1;
    }

    /**
     * Calculates the sum of the squares of the digits of a given number.
     * This generates the "next" number in the happy number sequence.
     *
     * @param num The input number.
     * @return The sum of the squares of its digits.
     */
    private int getNext(int num) {
        int totalSum = 0;
        while (num > 0) {
            int digit = num % 10;
            num = num / 10;
            totalSum += digit * digit;
        }
        return totalSum;
    }
}